ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,483 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+                    QuotaBlock->QuotaEntry[QuotaType].Limit, Amount);
+            return STATUS_PAGEFILE_QUOTA_EXCEEDED;
+        }
+
+        /* Guard ourselves quota in a spin lock */
+        KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
+
+        /*
+         * This is not a page file charge. What we can do at best
+         * in this scenario is to attempt to raise (expand) the
+         * quota limit charges of the block.
+         */
+        if (!MmRaisePoolQuota(QuotaType,
+                              QuotaBlock->QuotaEntry[QuotaType].Limit,
+                              &UpdatedLimit))
+        {
+            /*
+             * We can't? It could be that we must free
+             * up some resources in order to raise the
+             * limit, which in that case we must return
+             * the excess of quota that hasn't been
+             * returned. If we haven't returned anything
+             * then what we're doing here is futile.
+             * Bail out...
+             */
+            PspReturnExcessQuotas(QuotaType, &ReturnedQuotas);
+            if (ReturnedQuotas == 0)
+            {
+                DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Failed to free some resources in order to raise quota limits...\n");
+                KeReleaseSpinLock(&PspQuotaLock, OldIrql);
+                return STATUS_QUOTA_EXCEEDED;
+            }
+
+            /* Try to raise the quota limits again */
+            MmRaisePoolQuota(QuotaType,
+                             QuotaBlock->QuotaEntry[QuotaType].Limit,
+                             &UpdatedLimit);
+        }
+
+        /* Release the lock */
         KeReleaseSpinLock(&PspQuotaLock, OldIrql);
-        return STATUS_SUCCESS;
-        //return STATUS_QUOTA_EXCEEDED; /* caller raises the exception */
+
+        /* Enforce a new raised limit */
+        FinalLimit = QuotaBlock->QuotaEntry[QuotaType].Limit - UpdatedLimit;
+        InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit,
+                                        FinalLimit,
+                                        QuotaBlock->QuotaEntry[QuotaType].Limit);
+
+        /*
+         * Now determine if the current usage and the
+         * amounting by the caller still exceeds the
+         * quota limit of the process. If it's still
+         * over the limit then there's nothing we can
+         * do, so fail.
+         */
+        if (QuotaBlock->QuotaEntry[QuotaType].Usage + Amount >
+            QuotaBlock->QuotaEntry[QuotaType].Limit)
+        {
+            DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Quota amount exceeds the limit (limit -- %lu || amount -- %lu)\n",
+                    QuotaBlock->QuotaEntry[QuotaType].Limit, Amount);
+            return STATUS_QUOTA_EXCEEDED;
+        }
     }
 
-    InterlockedExchangeAdd((LONG*)&Process->QuotaUsage[QuotaType], Amount);
+    /* Update the quota usage */
+    InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Usage,
```

Comment:
As a rule, `InterlockedCompareExchange` isn't really useful if you're not looking at its return value.
I think the logic here needs to look something like
```
while (1)
{
    Usage = QuotaEntry->Usage; // fetch once
    Limit = QuotaEntry->Limit;

    if (Usage + Amount > Limit)
    {
        // fail or acquire more quota
    }
    if (InterlockedCompareExchange(&QuotaEntry->Usage, Usage + Amount, Usage) == Usage)
    {
        // CompareExchange succeeded
        break;
    }
    // somebody updated Usage while we were doing our calculations. Start over
}
```